### ISSUE TITLE
Globals have their own microtask queue that is separate from script thread's queue

### ISSUE THREAD

**Issue by jdm** [servo/servo#20908]
GlobalScope has a microtask queue that is used to support promises in worker runtimes. This is used from places like AutoEntryScript, which directly invokes `GlobalScope::perform_a_microtask_checkpoint`. Meanwhile, ScriptThread has its own `perform_a_microtask_checkpoint` which ignores the microtasks of any globals associated with that thread and uses its own internal microtask queue, which is manipulated by code like MutationObserver. We need to move the queue out of GlobalScope and into worker-specific subtypes, and ensure that GlobalScope delegates to ScriptThread or the subtypes as necessary.

**Comment by gterzian** [comment]
As part of that refactoring, we could maybe add a `perform_a_microtask_checkpoint` method to https://github.com/servo/servo/blob/347d8bdf727d6b43ba7f94b744c9115bc6226aff/components/script/dom/abstractworkerglobalscope.rs#L83

**Comment by jdm** [comment]
Huh, I wonder if this is related to the surprising behaviour that @KiChjang saw in https://github.com/servo/servo/pull/21456.

**Comment by gterzian** [comment]
> Huh, I wonder if this is related to the surprising behaviour that @KiChjang saw in #21456.

Since the parser calls the microtask checkpoints of globals, and there can be several globals in a script-thread, yes I think it could lead to problems. Also the spec mentions "Each event loop has a microtask queue", so we defenitely only want one per script-thread(and per worker thread).

**Comment by KiChjang** [comment]
This sounds like something I should look into.

**Comment by KiChjang** [comment]
@jdm I'm actually quite puzzled when you say that ScriptThread and GlobalScope use their own internal microtask queue, because the queue is currently stored as an `Rc<MicrotaskQueue>` everywhere and they are essentially shared between the script thread and the global scope.

**Comment by jdm** [comment]
I agree. I think I must have misread the code, since https://github.com/servo/servo/commit/f9034740f17f7b1eb7e5b5b9e2fca43b63d12a4b is quite a bit older than this issue and adds pretty clear documentation. However, reading through the code, I think I do see a problem - [the checkpoint code](https://github.com/servo/servo/blob/aa752e4d45ffad7f336850888ffe3e79fd1d8a08/components/script/microtask.rs#L94-L96) relies on the provided set of targets to decide whether to run promise job microtasks or silently discard them. When ScriptThread::perform_a_microtask is executed, the [full set](https://github.com/servo/servo/blob/aa752e4d45ffad7f336850888ffe3e79fd1d8a08/components/script/script_thread.rs#L3651-L3663) of same-origin globals are available. However, if the Window global's microtask check point is invoked instead, [all promise reactions](https://github.com/servo/servo/blob/aa752e4d45ffad7f336850888ffe3e79fd1d8a08/components/script/dom/globalscope.rs#L685) will end up using that global as the target, regardless of the pipeline id associated with that job.

To avoid this, I think we should make GlobalScope's microtask checkpoint API delegate to the ScriptThread one if we are a Window global.

**Comment by humancalico** [comment]
@jdm Can I work on this?

**Comment by jdm** [comment]
Of course!

**Comment by humancalico** [comment]
@jdm What do mean when you say if we are a window global and how would I check that?

**Comment by jdm** [comment]
@humancalico [This](https://github.com/servo/servo/blob/master/components/script/dom/globalscope.rs#L910-L915) is an example of some code that casts a GlobalScope into a Window. You can also use `some_global.is::<Window>()` for a simple yes/no if that is all that's necessary.

**Comment by jdm** [comment]
More specifically, GlobalScope is a base class of various kinds of globals - Window, WorkerGlobalScope, WorkletGlobalScope, and others. These correspond to the global object of various JS environments - page window objects, dedicated workers, and worklets.

**Comment by Gae24** [comment]
Should `MicrotaskQueue` still be behind an Rc in the `WorkerGlobalScope` case?

**Comment by jdm** [comment]
If it doesn't need to use an Rc then we should avoid it. I don't know the answer without doing some digging, though.

**Comment by Gae24** [comment]
It shouldn't since a worker agent has a single `WorkerGlobalScope`, while multiple windows can be owned by the same agent.
Actually, I should check if a new microtask queue should be created instead of cloning the one that gets passed down.
